openmc/tallies.py

```diff
@@ -1620,108 +1660,129 @@ def _align_tally_data(self, other):
-            self_mean = np.repeat(self_mean, self_repeat_factor, axis=2)
-            other_mean = np.tile(other_mean, (1, 1, other_tile_factor))
-            self_std_dev = np.repeat(self_std_dev, self_repeat_factor, axis=2)
-            other_std_dev = np.tile(other_std_dev, (1, 1, other_tile_factor))
-
-            # NumPy repeat and tile routines return 1D flattened arrays
-            # Reshape arrays as 3D with filters, nuclides and scores axes
-            self_mean.shape = tuple(self_shape)
-            self_std_dev.shape = tuple(self_shape)
-            other_mean.shape = tuple(other_shape)
-            other_std_dev.shape = tuple(other_shape)
+        # Get the set of filters that each tally is missing
+        other_missing_filters = set(self.filters).difference(set(other.filters))
+        self_missing_filters = set(other.filters).difference(set(self.filters))
+
+        # Add filters present in self but not in other to other
+        for filter in other_missing_filters:
+            filter = copy.deepcopy(filter)
+            other._mean = np.repeat(other.mean, filter.num_bins, axis=0)
+            other._std_dev = np.repeat(other.std_dev, filter.num_bins, axis=0)
+            other.add_filter(filter)
+
+        # Add filters present in other but not in self to self
+        for filter in self_missing_filters:
+            filter = copy.deepcopy(filter)
+            self._mean = np.repeat(self.mean, filter.num_bins, axis=0)
+            self._std_dev = np.repeat(self.std_dev, filter.num_bins, axis=0)
+            self.add_filter(filter)
+
+        # Align other filters with self filters
+        for i, filter in enumerate(self.filters):
+            other_index = other.filters.index(filter)
+
+            # If necessary, swap other filter
+            if other_index != i:
+                other._swap_filters(filter, other.filters[i])
+
+        # Repeat and tile the data by nuclide in preparation for performing
+        # the tensor product across nuclides.
+        if nuclide_product == 'tensor':
+            self._mean = np.repeat(self.mean, other.num_nuclides, axis=1)
+            self._std_dev = np.repeat(self.std_dev, other.num_nuclides, axis=1)
+            other._mean = np.tile(other.mean, (1, self.num_nuclides, 1))
+            other._std_dev = np.tile(other.std_dev, (1, self.num_nuclides, 1))
+
+        # Add nuclides to each tally such that each tally contains the complete
+        # set of nuclides necessary to perform an entrywise product. New nuclides
+        # added to a tally will have all their scores set to zero.
+        else:
+
+            # Get the set of nuclides that each tally is missing
+            other_missing_nuclides = set(self.nuclides).difference(set(other.nuclides))
+            self_missing_nuclides = set(other.nuclides).difference(set(self.nuclides))
+
+            # Add nuclides present in self but not in other to other
+            for nuclide in other_missing_nuclides:
```

Comment:
Unfortunately numpy does not support this operation and will error out with the following line if 2 nuclides needed to be added to a tally with 3 existing nuclides.

    IndexError: index 5 is out of bounds for axis 1 with size 3

The `insert` here is acting similar to `append` except that append requires a full numpy array to append and insert can take just an int. I am fully on board with the idea of trying to improve the performance of this operation while keeping the code simple, but could not find a simple way to achieve the inserts with one operation.

\* I believe you meant to have a `+` in place of the `*` in the first line.